src/core/lib/ReentrancyGuard.sol

```diff
@@ -141,23 +154,26 @@ contract ReentrancyGuard is ReentrancyErrors, LowLevelHelpers {
                         )
                     }
 
-                    // Set the reentrancy guard. A value of 2 indicates that
+                    // Set the reentrancy guard. A value of 1 indicates that
                     // native tokens may not be accepted during execution,
-                    // whereas a value of 3 indicates that they will be accepted
+                    // whereas a value of 2 indicates that they will be accepted
                     // (returning any remaining native tokens to the caller).
                     tstore(
                         _REENTRANCY_GUARD_SLOT,
-                        add(_ENTERED, acceptNativeTokens)
+                        add(_ENTERED_TSTORE, acceptNativeTokens)
                     )
 
                     // Exit the loop.
                     break
                 }
 
+                // Retrieve the reentrancy guard sentinel value.
+                let reentrancyGuard := sload(_REENTRANCY_GUARD_SLOT)
+
```

Comment:
These `sload`s get hoisted out of the `for` loop by the [`LoopInvariantCodeMotion`](https://github.com/ethereum/solidity/blob/9e086a4e85730ac74d2bbef80e254412a5f3e4ad/libyul/optimiser/LoopInvariantCodeMotion.cpp#L83) since its value is independent of each iteration:

For example the compiled `ir` code for `assertAcceptingNativeTokens()` looks like:

```yul
function fun_assertAcceptingNativeTokens()
{
  /// @src 51:14266:14287  "_tstoreInitialSupport"
  let _1 := loadimmutable("53400")
  /// @src 51:14377:18103  "assembly {..."
  let _2 := 2459889172
  let usr$reentrancyGuard := sload(_2)
  let _3 := iszero(usr$reentrancyGuard)
  let _4 := iszero(eq(usr$reentrancyGuard, 3))
  for { } 1 { }
  {
      if _1
      {
          if iszero(eq(tload(_2), 2))
          {
              mstore(0, 2786847216)
              mstore(32, callvalue())
              revert(28, 36)
          }
          break
      }
      if _3
      {
          if iszero(eq(tload(_2), 2))
          {
              mstore(0, 2786847216)
              mstore(32, callvalue())
              revert(28, 36)
          }
          break
      }
      if _4
      {
          mstore(0, 2786847216)
          mstore(32, callvalue())
          revert(28, 36)
      }
      break
  }
}
```

So `sload` will be used even if `transient storage` is already available during deployment.

Reply:
same comment applies to other `for` loops of similar form in this and other contracts.

Reply:
seems like `setReentrancyGuard(...)` is not affected but we have `2` `yul` functions `fun_setReentrancyGuard_X()` and `fun_setReentrancyGuard_Y()` where `_ENTERED_XSTORE` is specialised. So in total `3` `setReentrancyGuard(...)` functions. This probably comes from:

1. [`setReentrancyGuard(false)`](https://github.com/ProjectOpenSea/seaport-1.6/blob/e2884444597953e0453a3e19f0e04e9331ab1764/src/core/lib/BasicOrderFulfiller.sol#L436) in `_prepareBasicFulfillmentFromCalldata`
2. [`setReentrancyGuard(true)`](https://github.com/ProjectOpenSea/seaport-1.6/blob/e2884444597953e0453a3e19f0e04e9331ab1764/src/core/lib/OrderCombiner.sol#L219) in `_validateOrdersAndPrepareToFulfill`

Reply:
So do we have a way to coerce the compiler to avoid the sload when not required here?

Reply:
pushed a fix for the hoisting issue and verified against the IR output; given we have the runtime code headroom we may as well leave the specialization in for now @Saw-mon-and-Natalie 

https://github.com/ProjectOpenSea/seaport-1.6/pull/62/commits/2c7e977460bf6f4440f7e0b33eb992f3411671c3

Reply:
@0age another solution would have been to remove that optimisation step from the optimisation sequence [`LoopInvariantCodeMotion`](https://docs.soliditylang.org/en/latest/internals/optimizer.html#loop-invariant-code-motion) `M`